rotkehlchen/chain/ethereum/decoding/kyber/decoder.py

```diff
@@ -0,0 +1,125 @@
+KYBER_LEGACY_CONTRACT_UPGRADED = string_to_ethereum_address('0x9AAb3f75489902f3a48495025729a0AF77d4b11e')  # noqa: E501
+
+
+class KyberDecoder(DecoderInterface):
+    def __init__(  # pylint: disable=super-init-not-called
+            self,
+            ethereum_manager: 'EthereumManager',
+            base_tools: 'BaseDecoderTools',
+            msg_aggregator: MessagesAggregator,
+    ) -> None:
+        self.ethereum_manager = ethereum_manager
+        self.base = base_tools
+        self.msg_aggregator = msg_aggregator
+
+    def _decode_legacy_trade(  # pylint: disable=no-self-use
+        self,
+        tx_log: EthereumTxReceiptLog,
+        transaction: EthereumTransaction,  # pylint: disable=unused-argument
+        decoded_events: List[HistoryBaseEntry],
+        all_logs: List[EthereumTxReceiptLog],  # pylint: disable=unused-argument
+        action_items: Optional[List[ActionItem]],  # pylint: disable=unused-argument
+    ) -> Tuple[Optional[HistoryBaseEntry], Optional[ActionItem]]:
+        if tx_log.topics[0] == KYBER_TRADE_LEGACY:
+            return None, None
+
+        sender = hex_or_bytes_to_address(tx_log.topics[1])
+        source_token_address = hex_or_bytes_to_address(tx_log.data[:32])
+        destination_token_address = hex_or_bytes_to_address(tx_log.data[32:64])
+
+        source_token = ethaddress_to_asset(source_token_address)
+        if source_token is None:
+            return None, None
+        destination_token = ethaddress_to_asset(destination_token_address)
+        if destination_token is None:
+            return None, None
+
+        spent_amount_raw = hex_or_bytes_to_int(tx_log.data[64:96])
+        return_amount_raw = hex_or_bytes_to_int(tx_log.data[96:128])
+        spent_amount = asset_normalized_value(amount=spent_amount_raw, asset=source_token)
+        return_amount = asset_normalized_value(amount=return_amount_raw, asset=destination_token)
+
+        for event in decoded_events:
+            if event.event_type == HistoryEventType.SPEND and event.location_label == sender and event.asset == source_token and event.balance.amount == spent_amount:  # noqa: E501
+                event.event_type = HistoryEventType.TRADE
+                event.event_subtype = HistoryEventSubType.SPEND
+                event.counterparty = 'kyber legacy'
+                event.notes = f'Swap {event.balance.amount} {event.asset.symbol} in kyber'
+            elif event.event_type == HistoryEventType.RECEIVE and event.location_label == sender and event.balance.amount == return_amount and destination_token == event.asset:  # noqa: E501
+                event.event_type = HistoryEventType.TRADE
+                event.event_subtype = HistoryEventSubType.RECEIVE
+                event.counterparty = 'kyber legacy'
+                event.notes = f'Receive {event.balance.amount} {event.asset.symbol} from kyber swap'  # noqa: E501
+
+        return None, None
+
+    def _decode_legacy_upgraded_trade(  # pylint: disable=no-self-use
```

Comment:
Can't you create a common method called by oth of these? A lot of code from these 2 functions can be abstracted to a common one, right?